precompile/contracts/ibc/channel_test.go

```diff
@@ -0,0 +1,1366 @@
+			// run test for all types of ordering
+			for _, order := range []channeltypes.Order{channeltypes.UNORDERED, channeltypes.ORDERED} {
+
+				statedb, _ := state.New(common.Hash{}, state.NewDatabase(rawdb.NewMemoryDatabase()), nil)
+				statedb.Finalise(true)
+				vmctx := vm.BlockContext{
+					CanTransfer: func(vm.StateDB, common.Address, *big.Int) bool { return true },
+					Transfer:    func(vm.StateDB, common.Address, common.Address, *big.Int) {},
+				}
+				vmenv := vm.NewEVM(vmctx, vm.TxContext{}, statedb, params.TestChainConfig, vm.Config{ExtraEips: []int{2200}})
+
+				suite.SetupTest() // reset
+				path = ibctesting.NewPath(suite.chainA, suite.chainB)
+				path.EndpointA.ChannelConfig.Order = order
+				path.EndpointB.ChannelConfig.Order = order
+				tc.malleate()
+
+				counterparty := channeltypes.NewCounterparty(ibctesting.MockPort, ibctesting.FirstChannelID)
+
+				var input []byte
+
+				portIDbyte := []byte(path.EndpointA.ChannelConfig.PortID)
+				portIDLen := make([]byte, 8)
+				binary.BigEndian.PutUint64(portIDLen, uint64(len(portIDbyte)))
+
+				input = append(input, portIDLen...)
+				input = append(input, portIDbyte...)
+
+				channel := channeltypes.NewChannel(channeltypes.INIT, order, counterparty, []string{path.EndpointB.ConnectionID}, path.EndpointA.ChannelConfig.Version)
+
+				channelByte, _ := marshaler.Marshal(&channel)
+				channelLen := make([]byte, 8)
+				binary.BigEndian.PutUint64(channelLen, uint64(len(channelByte)))
+
+				input = append(input, channelLen...)
+				input = append(input, channelByte...)
+
+				input = append(getChanOpenInitSignature, input...)
+				admin := allowlist.TestAdminAddr
+				enableds := allowlist.TestEnabledAddr
+
+				allowlist.SetAllowListRole(vmenv.StateDB, ContractAddress, admin, allowlist.AdminRole)
+
+				contract := createIbcGoPrecompile()
+
+				suppliedGas := uint64(10000000)
+
+				connection, _ := suite.chainA.App.GetIBCKeeper().ConnectionKeeper.GetConnection(suite.chainA.GetContext(), path.EndpointA.ConnectionID)
+				connectionByte := marshaler.MustMarshal(&connection)
+				connectionsPath := fmt.Sprintf("connections/%s", path.EndpointA.ConnectionID)
+				vmenv.GetStateDB().SetPrecompileState(common.BytesToAddress([]byte(connectionsPath)), connectionByte)
+
+				cs, _ := suite.chainA.App.GetIBCKeeper().ClientKeeper.GetClientState(suite.chainA.GetContext(), path.EndpointA.ClientID)
+				cStore := suite.chainA.App.GetIBCKeeper().ClientKeeper.ClientStore(suite.chainA.GetContext(), path.EndpointA.ClientID)
+
+				if cs != nil {
+					clientState := cs.(*ibctm.ClientState)
+					bz := cStore.Get([]byte(fmt.Sprintf("consensusStates/%s", cs.GetLatestHeight())))
+					consensusState := clienttypes.MustUnmarshalConsensusState(marshaler, bz)
+					clientStateByte := clienttypes.MustMarshalClientState(marshaler, cs)
+
+					clientStatePath := fmt.Sprintf("clients/%s/clientState", connection.GetClientID())
+					vmenv.StateDB.SetPrecompileState(
```

Comment:
Comments on why SetPrecompileState() is called would be helpful. I see that the contract's Run() function uses the value, but the reason for setting/checking it should be commented somewhere I think.

Reply:
we create client state mock and store it into database. Then recompiles read it from database and use it